maintenance: implement PD maintenance endpoints to serialize TiKV maintenance operations

HaoW30 opened the pull request:
<!--

Thank you for working on PD! Please read PD's [CONTRIBUTING](https://github.com/tikv/pd/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?


<!--

Please create an issue first to describe the problem.
There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
For more info, check https://github.com/tikv/pd/blob/master/CONTRIBUTING.md#linking-issues.

-->
Issue Number: Close https://github.com/tikv/pd/issues/9477
### What is changed and how does it work?
  This commit implements the maintenance endpoints feature as specified in [RFC-0118](https://github.com/tikv/rfcs/blob/master/text/0118-pd-maintenance-endpoints.md)
  to prevent Raft quorum loss during TiKV maintenance operations.

  Key features:
  - Add atomic maintenance task storage operations using etcd transactions
  - Implement four HTTP endpoints: PUT, GET, GET by type, DELETE
  - Add pdctl CLI commands for maintenance task management
  - Integrate Prometheus metrics for monitoring
  - Ensure single maintenance task constraint with proper conflict handling<!--


### Check List

<!-- Remove the items that are not applicable. -->

Tests

<!-- At least one of these tests must be included. -->

- [x] Unit test

- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- [x]  Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

<!--

A bugfix or a new feature needs a release note. If there is no need to give a release note, just leave it with the `None`.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

-->

```release-note
None.
```

Commit:
maintenance: implement PD maintenance endpoints to serialize TiKV maintenance operations

This commit implements the maintenance endpoints feature as specified in RFC-0118
(https://github.com/tikv/rfcs/blob/master/text/0118-pd-maintenance-endpoints.md)
to prevent Raft quorum loss during TiKV maintenance operations.

Key features:
- Add atomic maintenance task storage operations using etcd transactions
- Implement four HTTP endpoints: PUT, GET, GET by type, DELETE
- Add pdctl CLI commands for maintenance task management
- Integrate Prometheus metrics for monitoring
- Ensure single maintenance task constraint with proper conflict handling

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

HaoW30 commented:
Will update docs onece this PR is approved

rleungx commented:
/ok-to-test

bufferflies commented on pkg/storage/endpoint/maintenance.go:
```diff
@@ -0,0 +1,239 @@
+	// TryStartMaintenanceTaskAtomic tries to start a maintenance task atomically.
+	TryStartMaintenanceTaskAtomic(ctx context.Context, task *MaintenanceTask) (bool, *MaintenanceTask, error)
+	// TryEndMaintenanceTaskAtomic tries to end a maintenance task atomically.
+	TryEndMaintenanceTaskAtomic(ctx context.Context, taskType, taskID string) (bool, *MaintenanceTask, error)
```
How about a unified TryDeleteTaskAtomic as the same as the client?

HaoW30 replied:
Renamed

SerjKol80 commented on pkg/storage/endpoint/maintenance.go:
```diff
@@ -0,0 +1,239 @@
+
+	// Else: get all tasks to see what's currently running
+	getAllOp := kv.RawTxnOp{
+		Key:    keypath.MaintenanceTaskPathPrefix(),          // Start key
+		OpType: kv.RawTxnOpGetRange,                          // Get range of keys
+		EndKey: keypath.MaintenanceTaskPathPrefix() + "\xff", // End key (prefix + max byte)
```
Will this skip all keys with prefix ````keypath.MaintenanceTaskPathPrefix() + "\xff"````?
It looks like if you set '0x00' in EndKey, the Prefix Get will be used instead of Raneg Get: https://github.com/tikv/pd/blob/f1b0c68e817505930e21a6ffb3c0022292a3c767/pkg/storage/kv/etcd_kv.go#L341-L346

But to be honest, that is pretty tricky and not readable. My suggestion is while you are here just add support of ````RawTxnOpGetPrefix````in this const: https://github.com/tikv/pd/blob/f1b0c68e817505930e21a6ffb3c0022292a3c767/pkg/storage/kv/kv.go#L37-L42

HaoW30 replied:
I think it should be fine, it only leaves out the keys that's lexicographically equal to or greater than prefix + "\xff".

\xff is not a printable ASCII char. 
In UTF-8, "\xff" is not valid on its own, but Go will treat it as a single byte in a string. So in this context, it’s safe and intentional.

Only if you actually have a key that starts with prefix + "\xff", which is unlikely, unless you explicitly used that value as part of a key.
Normally, all your keys will fall under prefix + [0x00 - 0xFE].

SerjKol80 commented on pkg/storage/endpoint/maintenance.go:
```diff
@@ -0,0 +1,239 @@
+			}
+		}
+
+		// Assert that there should be only one task
+		if taskCount > 1 {
+			// This shouldn't happen - multiple tasks running simultaneously
+			return false, nil, fmt.Errorf("multiple maintenance tasks running simultaneously: found %d tasks", taskCount)
+		}
```
What do you think about printing all found tasks? More context will help in debugging if we ever face this issue.

SerjKol80 reviewed (commented):
How do different components (ie. k8s-operator) suppose to distinguish hit own operations from others'? Is it supposed that it needs to encode his ID into taskID?

bufferflies commented on pkg/storage/endpoint/maintenance.go:
```diff
@@ -0,0 +1,239 @@
+	// LoadMaintenanceTask loads a maintenance task by type from storage.
+	LoadMaintenanceTask(taskType string) (*MaintenanceTask, error)
+	// LoadAllMaintenanceTasks loads all maintenance tasks from storage.
+	LoadAllMaintenanceTasks(f func(k, v string)) error
```
Why need to input one function? It's better to return all tasks and hide the function in the implementation.

bufferflies commented on tools/pd-ctl/pdctl/command/maintenance_command.go:
```diff
@@ -0,0 +1,186 @@
+// extractExistingTaskInfo tries to extract existing task information from error response
+func extractExistingTaskInfo(errorMsg string) string {
+	// Look for JSON content in the error message
+	start := strings.Index(errorMsg, "{")
```
It seems strange to decode the error message manually.

HaoW30 replied:
Yes, the reason I added this was to provide a more human-readable error message in the pd-ctl command response. That’s why I included decoding of the responses.

Let me know if you feel this is too cumbersome, I can update it to return the raw HTTP response directly as the pd-ctl result instead.

bufferflies replied:
I think it's better to return the raw http response directly

Commit:
Rename end task functions to delete task.

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

HaoW30 commented on pkg/utils/keypath/absolute_key_path.go:
```diff
@@ -58,6 +58,9 @@ const (
 	regionLabelPrefixFormat = "/pd/%d/region_label/"   // "/pd/{cluster_id}/region_label/"
 
+	// Maintenance task path format
+	maintenanceTaskPathFormat = "/pd/%d/maintenance/%s" // "/pd/{cluster_id}/maintenance/{task_type}"
```
@SerjKol80 I think this is a more general question for PingCap folks. 

From my understanding, this is related to multi-cluster support—allowing multiple clusters to share the same etcd infrastructure. In this context, the cluster ID acts like a namespace identifier to isolate data between clusters.
https://github.com/tikv/pd/blob/6bdfbd398bce2e0cf2a3753a180470d12737699d/tests/integrations/realcluster/cluster_id_test.go#L40-L63
It may also be required for BR and recovery operations.
https://github.com/tikv/pd/blob/6bdfbd398bce2e0cf2a3753a180470d12737699d/tools/pd-recover/main.go#L114-L163

HaoW30 commented on pkg/storage/endpoint/maintenance.go:
```diff
@@ -0,0 +1,239 @@
+
+// TryStartMaintenanceTaskAtomic tries to start a maintenance task atomically.
+// Returns (true, nil) if success, (false, existingTask, nil) if conflict, (false, nil, err) if error.
+func (se *StorageEndpoint) TryStartMaintenanceTaskAtomic(_ context.Context, task *MaintenanceTask) (bool, *MaintenanceTask, error) {
```
@bufferflies Good point. I actually considered both approaches during implementation. While using an in-memory lock combined with separate etcd operations can achieve similar results to an etcd transaction, I chose to go with the etcd transaction for the following reasons:

Built-in support: PD already provides robust support for etcd transactions through storage/kv/etcd_kv.go, so there's no need to implement anything new. We can simply build on top of the atomicity guarantees etcd already provides.
True atomicity: etcd ensures true atomicity, whereas the in-memory lock approach requires careful review of corner cases. It also removes the burden of checking for leadership and explicitly wrapping updates with LeaderTxn() to ensure they go through the PD leader.
Better extensibility: While in-memory locking might seem simpler at first—possibly saving one etcd write—it introduces extensibility issues. If we later expand support to multiple types of maintenance tasks and allow independent updates to each type, etcd transactions can elegantly handle this without requiring multiple locks for each type of task_type.
Future-proofing follower handling: According to the current design, maintenance task tracking serves as a system of record. There’s no fundamental reason this endpoint couldn’t be handled by followers in the future (via PD-Allow-Follower-Handle). Using etcd transactions aligns well with that potential evolution, making the transition easier and more robust.
Low performance sensitivity: Since maintenance task registration is expected to be a pretty low-frequency operation, I’m not particularly concerned about the performance difference between the two approaches.

That said, if you have strong reasons for preferring the in-memory lock solution, I’m happy to discuss further and explore alternatives.

bufferflies replied:
No, thank you for your analysis.

Commit:
explicitly allocate task on heap.

Make it easier for golang escape analysis to correctly allocate the task on heap to avoid data corruption

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

HaoW30 commented:
@bufferflies gentle ping... thanks!

bufferflies commented on pkg/storage/endpoint/maintenance.go:
```diff
@@ -0,0 +1,239 @@
+	rawTxn := se.CreateRawTxn()
+
+	// Maintenance lock key to enforce single task type constraint
+	lockKey := keypath.MaintenanceTaskPath("__maintenance_lock__")
```
What I mean is that you can define a constant, such as LockName = "__maintenance_lock__", and then use it here.

HaoW30 replied:
Done

HaoW30 commented:
> How do different components (ie. k8s-operator) suppose to distinguish hit own operations from others'? Is it supposed that it needs to encode his ID into taskID?

That’s entirely up to the orchestrator (k8s operator). Yes, I think ideally, the operator should include its identifier in the task type or task ID.

Commit:
Address PR comments

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

Commit:
clean up lock name

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

Commit:
add test case

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

bufferflies commented on server/apiv2/handlers/maintenance.go:
```diff
@@ -0,0 +1,203 @@
+// @Failure  404  {string}  string  "No maintenance task is running."
+// @Failure  500  {string}  string  "PD server failed to proceed the request."
+// @Router   /maintenance [get]
+func GetMaintenanceTask(c *gin.Context) {
```
it should be renamed as `GetMaintenanceTasks` if this returns all the task.

HaoW30 replied:
Updated

bufferflies commented on server/apiv2/handlers/maintenance.go:
```diff
@@ -0,0 +1,203 @@
+	c.JSON(http.StatusOK, "Maintenance task started successfully.")
+}
+
+// GetMaintenanceTask gets information about the ongoing maintenance task.
```
```suggestion
// GetMaintenanceTask gets information about the ongoing maintenance tasks.
```

bufferflies commented on server/apiv2/handlers/maintenance.go:
```diff
@@ -0,0 +1,203 @@
+
+// GetMaintenanceTask gets information about the ongoing maintenance task.
+// @Tags     maintenance
+// @Summary  Get information about the ongoing maintenance task.
```
```suggestion
// @Summary  Get information about the ongoing maintenance tasks.
```

HaoW30 replied:
Done

bufferflies commented on server/apiv2/handlers/maintenance.go:
```diff
@@ -0,0 +1,203 @@
+		Description:    description,
+	}
+
+	ok, existingTask, err := svr.GetStorage().TryStartMaintenanceTaskAtomic(c.Request.Context(), task)
```
Do we need to print some logs to debug  this inner status if something's wrong and it's helpful to record all the behavior

HaoW30 replied:
Added more log prints for TryStartMaintenanceTaskAtomic and TryDeleteMaintenanceTaskAtomic in pkg/storage/endpoint/maintenance.go

bufferflies commented on server/apiv2/handlers/maintenance.go:
```diff
@@ -0,0 +1,203 @@
+// @Failure  409  {object}  map[string]interface{}  "Task ID does not match the current task."
+// @Failure  500  {string}  string  "PD server failed to proceed the request."
+// @Router   /maintenance/{task_type}/{task_id} [delete]
+func DeleteMaintenanceTask(c *gin.Context) {
```
ditto

bufferflies commented on tools/pd-ctl/pdctl/command/maintenance_command.go:
```diff
@@ -0,0 +1,125 @@
+			}
+
+			// Print the raw response
+			cmd.Println(resp)
```
Does it need to decode the timestamp to a readable time string?

HaoW30 replied:
I’d like to keep the raw timestamp in the HTTP response, as it’s easier for the k8s operator to handle.
As for the pdctl command response, I think it’s fine to return the raw Unix epoch—it’s straightforward for a human user to decode.
If we were to decode the timestamp for the user, we’d be reverting to the old approach of parsing the HTTP response and handling all the potential error scenarios.

Commit:
Address PR comments
 *fix naming
 *enable more logs printing for the functions that do atomic etcd record updates

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

Commit:
Define maintenancePrefixURI for the maintenance endpoint URI prefix

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

HaoW30 commented on tools/pd-ctl/pdctl/command/maintenance_command.go:
```diff
@@ -0,0 +1,125 @@
+		Run: func(cmd *cobra.Command, args []string) {
+			taskType := args[0]
+			taskID := args[1]
+			path := fmt.Sprintf("pd/api/v2/maintenance/%s/%s", taskType, taskID)
```
Done

Commit:
fix build

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

Commit:
fix lint errors

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

Commit:
fix lint

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

Commit:
lint fix

Signed-off-by: h-wang4 <h.wang4@airbnb.com>

HaoW30 commented:
@okJiang could you please help to trigger the rerun of the ci tests again?

bufferflies commented:
/retest

HaoW30 commented:
/retest
